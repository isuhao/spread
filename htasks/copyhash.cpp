#include "copyhash.hpp"
#include <mangle/stream/servers/file_stream.hpp>
#include <mangle/stream/clients/copy_stream.hpp>

using namespace Spread;
using namespace Mangle::Stream;

// This just reads from a file and dumps the data into a stream
struct CopyTask : Jobify::Job
{
  CopyTask(const std::string &input, StreamPtr output)
    : in(input), out(output) {}

  void doJob()
  {
    setBusy("Copying " + in);
    CopyStream::copy(FileStream::Open(in), out);
    setDone();
  }

private:
  std::string in;
  StreamPtr out;
};

Jobify::Job* CopyHash::createJob()
{
  // Guess the hash from targets
  assert(outputs.size() > 0);
  const Hash &hash = outputs.begin()->first;
  return new CopyTask(source, getOutStream(hash));
}
