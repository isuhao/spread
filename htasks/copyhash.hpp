#ifndef __HASH_COPYTASK_HPP_
#define __HASH_COPYTASK_HPP_

#include "hashtask.hpp"

/* This class copy one file to a given output set. The hash is guessed
   from the first entry in the copy list.
 */

namespace Spread
{
  struct CopyHash : HashTask
  {
    std::string source;

  private:
    Jobify::Job *createJob();
  };
};

#endif
