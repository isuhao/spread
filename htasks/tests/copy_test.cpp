#include "copyhash.hpp"

#include <iostream>
using namespace std;
using namespace Spread;

int main()
{
  string source = "test1.zip";
  cout << "Copying " << source << " ...\n";

  Hash hash("OEa8aUtr8n77FuIrXefgKVpVRY66g2CULBYP4N9EuQIEBw");

  CopyHash dl;
  dl.source = source;
  dl.addOutput(hash, "_copy1.zip");
  dl.addOutput(hash, "_copy2.dat");
  dl.run(false);

  return 0;
}
