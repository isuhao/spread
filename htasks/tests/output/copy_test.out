Copying test1.zip ...
